blog/_src/posts/2018-10-05-defining-local-bindings-in-turnstile-languages.md

```diff
@@ -0,0 +1,623 @@
+   `type` [syntax class][19] is provided by Turnstile, and describes the surface
+   syntax of types (such as those created with `define-type-constructor`);
+   internal operations over types use the expanded version of the type, which is
+   accessed via the `norm` attribute.
+4. The chevron `≫` on the first line signifies that there is only one case in
+   this type rule. Some rules, which we will see later, use multiple
+   cases to check different kinds of uses.
+5. The body of the rule is a sequence of premises, that usually check and
+   analyze the types of sub-expressions, followed by a dashed line, and
+   then the conclusion, describing the output syntax and its type.
+6. Here, the single premise describes how to check the body of the function. The
+   context, which associates variables with types, goes to the left of the
+   turnstile (`⊢`). For each formal `x`, this lets us know what type `x` has
+   when we find a reference to it in `e`. In this rule, we are saying "while
+   checking the right-hand-side, assume `x`---which elaborates to
+   `x-`---has type `τ_in`, for each triple in the input syntax (signified by the
+   ellipses `...`)". More on the "renaming to `x-`" below.
+7. To the right of the turnstile, we write the expression we are checking, `e`,
+   and patterns `e-` and `τ_out` matching the elaboration of `e` and its type,
+   respectively.
+8. After the dashes comes the conclusion, which begins with `⊢`. The next part
+   specifies the elaboration of the term. Here, the meaning of the typed `λ` is
+   given in terms of Racket's [`#%plain-lambda`][5]. Turnstile uses the
+   convention of a `-` suffix for forms in the untyped/target language to avoid
+   conflicting names and confusion. Suffixed names are usually bound using
+   `postfix-in`, such as in `(require (postfix-in - racket/base))` to bind
+   `#%plain-lambda-`.
+9. Finally, we give the type of the term to the right of the `⇒`, referring to
+   variables bound in the premises.
+   
+#### Renaming Typed Variables
+   
+Turnstile lets the Racket expander take care of the details of variable scope,
+shadowing, etc. To associate identifier `x` with type `τ`, Turnstile binds `x`
+to a macro that knows `τ` when it expands. References to `x` now become
+references to that macro, and expanding them provides access to `τ`. Concretely,
+the code to implement this method looks something like this:
+
+```racket
+(let ([x- (assign-type (generate-temporary #'x) #'τ)])
+  (let-syntax ([x (make-rename-transformer x-)])
+    ... expand and check forms that may reference x ...))
+```
+
+The type `τ` is attached as [metadata][6] for a new identifier `x-`, which is
+what `x` will transform to at any reference site. In order for this to work,
+`x-` must be distinct from `x`. While it is possible to define `x` as a macro
+that expands to `x` plus type information, it's a bit tricky to retrieve type
+information from an infinite loop.
```

Comment:
zing!

Though, since its a joke, maybe put this "While ...." sentence in parentheses or prefix it with a ">" to blockquote it?